File: Heap/Heap01.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    priority_queue<int, vector<int>, greater<int>> pq;  // 우선순위 큐 int형 vector<int>구조로 오름차순 정리
    for(int i=0,j=scoville.size();i<j;i++){
        pq.push(scoville[i]);
    }
    while(pq.top()<K){
        if(pq.size()<2){
            answer=-1;
            break;
        }
        int first=pq.top();
        pq.pop();
        int second=pq.top();
        pq.pop();
        pq.push(first+(second*2));
        answer++;
    }

    return answer;
}

File: Heap/Heap01.py
import heapq

def solution(scoville, K):
    min = 0
    answer = 0

    # min heap 구조로 만들어줌
    heapq.heapify(scoville)

    # heap의 가장 작은 노드가 K보다 작지 않을 때 까지 섞음
    while scoville[0] < K:

        # 원소가 하나밖에 남지 않았는데 그 값이 K보다 작으면 return -1
        if len(scoville) == 1:
            return -1

        min = heapq.heappop(scoville)
        min += heapq.heappop(scoville) * 2
        heapq.heappush(scoville, min)
        answer += 1

    return answer

scoville = [1,2,0,0,0,0]
K = 7
print(solution(scoville, K))